### PR TITLE
INT-2277: Only update event bindings that change

### PR DESCRIPTION
**Pull request by tiny-james**
This issue:
https://github.com/tinymce/tinymce-react/issues/187

Identified that we were regenerating event bindings on every prop change not just changes to the event binding. This PR aims to address that issue by calculating which event props were removed, changed or added and only operating on those.

Also updated the tinymce types to latest to address issue https://github.com/tinymce/tinymce-react/issues/190

### CODE REVIEW

File: src/main/ts/Utils.ts [hunk 1]
```diff
@@ -17,33 +16,35 @@ const isEventProp = (name: string): name is keyof IEventPropTypes => {
   return name in eventPropTypes;
 };
 
-interface EventHandlerSet {
-  handler: EventHandler<unknown>;
-  eventName: string;
+const eventAttrToEventName = <T extends string>(attrName: `on${string & T}`): T => {
```
Comment (techtangents) [inline src/main/ts/Utils.ts:19]: Do you need `string & T` here? Would `T` suffice?

**Comment by tiny-james** [comment]
It looks like LGTM doesn't understand typescript template literal types yet...

File: src/main/ts/Utils.ts (hunk 1 above)
Comment (tiny-james) [reply src/main/ts/Utils.ts:19]: It's what they do in the examples but I didn't try it.

**Review by techtangents — APPROVED** [review]
I'm no expert, but it looks like what you've described, and the code reads fine.

How hard would it be to add a test here?

File: src/main/ts/Utils.ts (hunk 1 above)
Comment (tiny-james) [reply src/main/ts/Utils.ts]: I tried it without the `string &` and it seems to still work so I removed it.

**Comment by tiny-james** [comment]
@techtangents tricky to test because the state we need to check is inside TinyMCE. The problem is that we want to know about the calls to `tinymce.on` or `tinymce.off` functions. to make sure we're not doing unneeded rebinding.

Maybe I could change the method and make it take in `on` and `off` as separate functions rather than just taking TinyMCE as a whole. Then the function could be tested in isolation with mock `on` and `off` functions. What do you think?

**Comment by techtangents** [comment]
@EphoxJames The mock functions could work. It's worth trying. Hopefully it doesn't make the code messy as a result.

What about just a test that the correct things are bound before and after? Or does that suffer the same problem?

**Review by jscasca — APPROVED** [review]

**Review by SimonFc — APPROVED** [review]

**Pull request merged**